h2o-core/src/main/java/water/api/RemoveAllHandler.java

```diff
@@ -23,14 +20,47 @@ public RemoveAllV3 remove(int version, RemoveAllV3 u) {
       RapidsHandler.SESSIONS.clear();
     }
     fs.blockForPending();
+
+    if (u.retained_keys != null && u.retained_keys.length != 0) {
+      retainKeys(u.retained_keys);
+    } else {
+      clearAll();
+    }
+    Log.info("Finished removing objects");
+    return u;
+  }
+
+
+  private void clearAll() {
+    Log.info("Removing all objects");
     // Bulk brainless key removal.  Completely wipes all Keys without regard.
-    new MRTask(H2O.MIN_HI_PRIORITY){
-      @Override public void setupLocal() {  H2O.raw_clear();  water.fvec.Vec.ESPC.clear(); }
+    new MRTask(H2O.MIN_HI_PRIORITY) {
+      @Override
+      public void setupLocal() {
+        H2O.raw_clear();
+        water.fvec.Vec.ESPC.clear();
+      }
     }.doAllNodes();
     // Wipe the backing store without regard as well
     H2O.getPM().getIce().cleanUp();
     H2O.updateNotIdle();
-    Log.info("Finished removing objects");
-    return u;
   }
-}
\ No newline at end of file
+
+  private void retainKeys(final KeyV3[] retained_keys) {
+    Log.info(String.format("Removing all objects, except for %d provided key(s)", retained_keys.length));
+    final Key[] retainedKeys;
+    if (retained_keys == null) {
+      retainedKeys = new Key[0];
+    } else {
+      retainedKeys = new Key[retained_keys.length];
+      for (int i = 0; i < retainedKeys.length; i++) {
+        if (retained_keys[i] == null) continue; // Protection against null keys from the client - ignored
```

Comment:
It would probably be better to avoid passing null to ClearDKVTask

Reply:
Retaining null results in an effect no-op. But the user won't know. The documentation says nulls are silently ignored.

I'll add a verification mechanism.

Reply:
Added. Also please note the DKVCleaner itself does not validate if the key names are actually valid. Non-valid/not existing keys will simply be ignored, as the algorithm iterates keys in DKV and tries to match them against a set with keys to retain.